File: geophone-python/REST/get_phone_info_last_first_rest.py
'''
Service Objects, Inc.

This module provides the get_phone_info_last_first function to validate and retrieve phone number
information using the Service Objects GP API's GetPhoneInfoLastFirst endpoint. It handles live/trial
endpoints, fallback logic, and JSON parsing.

Functions:
    get_phone_info_last_first(phone_number: str, license_key: str, is_live: bool, timeout_seconds: int) -> dict:
'''

import requests  # HTTP client for RESTful API calls
from gp_response import GPResponse, Provider, Contact, Error

# Endpoint URLs for GP GetPhoneInfoLastFirst service
primary_url = 'https://sws.serviceobjects.com/GP/api.svc/json/GetPhoneInfoLastFirst?'
backup_url = 'https://swsbackup.serviceobjects.com/GP/api.svc/json/GetPhoneInfoLastFirst?'
trial_url = 'https://trial.serviceobjects.com/GP/api.svc/json/GetPhoneInfoLastFirst?'


def get_phone_info_last_first(phone_number: str, license_key: str, is_live: bool = True) -> GPResponse:
    """
    Call GP GetPhoneInfoLastFirst API to retrieve phone number information.

    Parameters:
        phone_number (str): Phone number to look up (e.g., '805-963-1700'). Required.
        license_key (str): Service Objects license key. Required.
        is_live (bool): True for production endpoints, False for trial URL. Defaults to True.
        timeout_seconds (int): Timeout for API calls in seconds. Defaults to 30.

    Returns:
        dict: Parsed JSON response with phone information or error details.
    """
    # Prepare query parameters for GP API
    params = {
        'PhoneNumber': phone_number,
        'LicenseKey': license_key
    }

    # Select the base URL: production vs trial
    url = primary_url if is_live else trial_url

    # Attempt primary (or trial) endpoint first
    try:
        response = requests.get(url, params=params, timeout=10)
        response.raise_for_status()
        data = response.json()

        # If API returned an error in JSON payload, trigger fallback
        error = getattr(response, 'Error', None)
        if not (error is None or getattr(error, 'Number', None) != "4"):
            if is_live:
                # Try backup URL when live
                response = requests.get(backup_url, params=params, timeout=10)
                response.raise_for_status()
                data = response.json()
                # If still error, propagate exception
                if 'Error' in data:
                    raise RuntimeError(f"GP service error: {data['Error']}")
            else:
                # Trial mode should not fallback; error is terminal
                raise RuntimeError(f"GP trial error: {data['Error']}")

        # Convert JSON response to GPResponse for structured access
        error = Error(**data.get('Error', {})) if data.get('Error') else None
        providers = [Provider(**prov) for prov in data.get('Providers', [])] if data.get('Providers') else None
        contacts = [Contact(**cont) for cont in data.get('Contacts', [])] if data.get('Contacts') else None
        response = GPResponse(providers=providers, contacts=contacts, error=error)

        return response

    except requests.RequestException as req_exc:
        # Network or HTTP-level error occurred
        if is_live:
            try:
                # Fallback to backup URL on network failure
                response = requests.get(backup_url, params=params, timeout=10)
                response.raise_for_status()
                data = response.json()
                if 'Error' in data and data['Error'].get('Number') == "4":
                    raise RuntimeError(f"GP backup error: {data['Error']}") from req_exc

                # Convert JSON response to GPResponse for structured access
                error = Error(**data.get('Error', {})) if data.get('Error') else None
                providers = [Provider(**prov) for prov in data.get('Providers', [])] if data.get('Providers') else None
                contacts = [Contact(**cont) for cont in data.get('Contacts', [])] if data.get('Contacts') else None
                response = GPResponse(providers=providers, contacts=contacts, error=error)

                return response

            except Exception as backup_exc:
                # Both primary and backup failed; escalate
                raise RuntimeError("GP service unreachable on both endpoints") from backup_exc
        else:
            # In trial mode, propagate the network exception
            raise RuntimeError(f"GP trial error: {str(req_exc)}") from req_exc

File: geophone-nodejs/src/REST/get-phone-info-last-first-rest.js
import axios from 'axios';
import querystring from 'querystring';
import { GPResponse } from './gp-response.js';

/**
* @constant
* @type {string}
* @description The base URL for the live ServiceObjects GeoPhone API service.
*/
const LiveBaseUrl = 'https://sws.serviceobjects.com/GP/api.svc/';

/**
* @constant
* @type {string}
* @description The base URL for the backup ServiceObjects GeoPhone API service.
*/
const BackupBaseUrl = 'https://swsbackup.serviceobjects.com/GP/api.svc/';

/**
* @constant
* @type {string}
* @description The base URL for the trial ServiceObjects GeoPhone API service.
*/
const TrialBaseUrl = 'https://trial.serviceobjects.com/GP/api.svc/';

/**
* Checks if a response from the API is valid by verifying that it either has no Error object
* or the Error.Number is not equal to '4'.
* @param {Object} response - The API response object to validate.
* @returns {boolean} True if the response is valid, false otherwise.
*/
const isValid = (response) => !response?.Error || response.Error.Number !== '4';

/**
* Constructs a full URL for the GetPhoneInfoLastFirst API endpoint by combining the base URL
* with query parameters derived from the input parameters.
* @param {string} phoneNumber - The phone number to query.
* @param {string} licenseKey - The license key for API authentication.
* @param {string} baseUrl - The base URL for the API service (live, backup, or trial).
* @returns {string} The constructed URL with query parameters.
*/
const buildUrl = (phoneNumber, licenseKey, baseUrl) =>
    `${baseUrl}json/GetPhoneInfoLastFirst?${querystring.stringify({ PhoneNumber: phoneNumber, LicenseKey: licenseKey })}`;

/**
* Performs an HTTP GET request to the specified URL with a given timeout.
* @param {string} url - The URL to send the GET request to.
* @param {number} timeoutSeconds - The timeout duration in seconds for the request.
* @returns {Promise<GPResponse>} A promise that resolves to a GPResponse object containing the API response data.
* @throws {Error} If the HTTP request fails, with a message detailing the error.
*/
const httpGet = async (url, timeoutSeconds) => {
    try {
        const response = await axios.get(url, { timeout: timeoutSeconds * 1000 });
        return new GPResponse(response.data);
    } catch (error) {
        throw new Error(`HTTP request failed: ${error.message}`);
    }
};

/**
* Provides functionality to call the ServiceObjects GeoPhone API's GetPhoneInfoLastFirst endpoint,
* retrieving phone-related information with fallback to a backup endpoint for reliability in live mode.
*/
const GetPhoneInfoLastFirstClient = {
    /**
    * Asynchronously invokes the GetPhoneInfoLastFirst API endpoint, attempting the primary endpoint
    * first and falling back to the backup if the response is invalid (Error.Number == "4") in live mode.
    * @param {string} phoneNumber - The phone number to query.
    * @param {string} licenseKey - The license key for API authentication.
    * @param {boolean} isLive - Whether to use live or trial endpoint.
    * @param {number} [timeoutSeconds=15] - The timeout duration in seconds for the request.
    * @returns {Promise<GPResponse>} A promise that resolves to a GPResponse object with provider and contact details or an error.
    * @throws {Error} If both primary and backup endpoints fail, with details of the failure.
    */
    async invokeAsync(phoneNumber, licenseKey, isLive, timeoutSeconds = 15) {
        const url = buildUrl(phoneNumber, licenseKey, isLive ? LiveBaseUrl : TrialBaseUrl);
        let response = await httpGet(url, timeoutSeconds);
    
        if (isLive && !isValid(response)) {
            const fallbackUrl = buildUrl(phoneNumber, licenseKey, BackupBaseUrl);
            const fallbackResponse = await httpGet(fallbackUrl, timeoutSeconds);
            return isValid(fallbackResponse) ? fallbackResponse : response;
        }
        return response;
    },

    /**
    * Synchronously invokes the GetPhoneInfoLastFirst API endpoint by wrapping the async call
    * and awaiting its result immediately.
    * @param {string} phoneNumber - The phone number to query.
    * @param {string} licenseKey - The license key for API authentication.
    * @param {boolean} isLive - Whether to use live or trial endpoint.
    * @param {number} [timeoutSeconds=15] - The timeout duration in seconds for the request.
    * @returns {GPResponse} A GPResponse object with provider and contact details or an error.
    * @throws {Error} If both primary and backup endpoints fail, with details of the failure.
    */
    invoke(phoneNumber, licenseKey, isLive, timeoutSeconds = 15) {
        return (async () => await this.invokeAsync(phoneNumber, licenseKey, isLive, timeoutSeconds))();
    },
};

export { GetPhoneInfoLastFirstClient, GPResponse };